File: Java/com/leetcode/maximumDepthOfBinaryTree/MaximumDepthOfBinaryTree.java
package com.leetcode.maximumDepthOfBinaryTree;

import com.leetcode.dataStructures.TreeNode;

public class MaximumDepthOfBinaryTree {
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        int leftMax = maxDepth(root.left);
        int rightMax = maxDepth(root.right);
        
        if (leftMax < rightMax) {
            return rightMax + 1;
        }
        return leftMax + 1;
    }
    
    /*
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        LinkedList<TreeNode> queue = new LinkedList<TreeNode>();
        TreeNode dummy = new TreeNode(-1);
        
        queue.add(root);
        queue.add(dummy);
        int length = 2;
        int depth = 1;
        while (length > 0) {
            TreeNode n = queue.pop();
            --length;
            
            if (n == dummy) {
                if (length > 0) {
                    ++depth;
                    queue.add(dummy);
                    ++length;
                }
            } else {
                if (n.left != null) {
                    queue.add(n.left);
                    ++length;
                }
                if (n.right != null) {
                    queue.add(n.right);
                    ++length;
                }
            }
        }
        
        return depth;
    }
     */
}

File: Python/maximum_depth_of_binary_tree/maximum_depth_of_binary_tree.py
class Solution(object):
    def maxDepth(self, root):
        max_depth = [0]
        self.helper(root, 0, max_depth)
        return max_depth

    def helper(self, root, curr, max_depth):
        if root is None:
            max_depth = max_depth \
                    if max_depth[0] > curr else [curr]
            return
        curr += 1
        self.helper(root.left, curr, max_depth)
        self.helper(root.right, curr, max_depth)
